test/integration/test_workflow_refactoring.py

```diff
@@ -50,7 +58,7 @@ def setUp(self):
     def test_basic_refactoring_types(self):
         self.workflow_populator.upload_yaml_workflow(REFACTORING_SIMPLE_TEST)
 
-        actions = [
+        actions: ActionsJson = [
```

Comment:
Do we need this (and similarly below)? Mypy should be able to figure out the type by itself, no?

Reply:
I believe so yes, this is the output without this:

```
test/integration/test_workflow_refactoring.py:80: error: Dict entry 1 has incompatible type
"str": "List[Dict[str, str]]"; expected "str": "str"  [dict-item]
                {"action_type": "update_creator", "creator": [{"class": "Person", "name...
                                                  ^
test/integration/test_workflow_refactoring.py:86: error: Dict entry 1 has incompatible type
"str": "Dict[str, str]"; expected "str": "str"  [dict-item]
            actions = [{"action_type": "update_report", "report": {"markdown": "my repo...
                                                        ^
test/integration/test_workflow_refactoring.py:92: error: Dict entry 1 has incompatible type
"str": "Dict[str, int]"; expected "str": "str"  [dict-item]
                {"action_type": "update_step_label", "step": {"order_index": 0}, "label...
                                                     ^
test/integration/test_workflow_refactoring.py:103: error: Dict entry 3 has incompatible type
"str": "Dict[str, str]"; expected "str": "str"  [dict-item]
                    "tool_state": {"parameter_type": "text"},
                    ^
test/integration/test_workflow_refactoring.py:104: error: Dict entry 4 has incompatible type
"str": "Dict[str, int]"; expected "str": "str"  [dict-item]
                    "position": {"left": 10, "top": 50},
                    ^
test/integration/test_workflow_refactoring.py:117: error: Dict entry 1 has incompatible type
"str": "Dict[str, str]"; expected "str": "str"  [dict-item]
                    "step": {"label": "new_param"},
                    ^
test/integration/test_workflow_refactoring.py:118: error: Dict entry 2 has incompatible type
"str": "Dict[str, int]"; expected "str": "str"  [dict-item]
                    "position_shift": {"left": 3, "top": 5},
                    ^
test/integration/test_workflow_refactoring.py:132: error: Dict entry 3 has incompatible type
"str": "bool"; expected "str": "str"  [dict-item]
                    "optional": True,
                    ^
test/integration/test_workflow_refactoring.py:133: error: Dict entry 4 has incompatible type
"str": "Dict[str, int]"; expected "str": "str"  [dict-item]
                    "position": {"top": 1, "left": 2},
                    ^
test/integration/test_workflow_refactoring.py:146: error: Dict entry 1 has incompatible type
"str": "Dict[str, str]"; expected "str": "str"  [dict-item]
                    "input": {"label": "first_cat", "input_name": "input1"},
                    ^
test/integration/test_workflow_refactoring.py:147: error: Dict entry 2 has incompatible type
"str": "Dict[str, str]"; expected "str": "str"  [dict-item]
                    "output": {"label": "new_label"},
                    ^
test/integration/test_workflow_refactoring.py:156: error: Dict entry 1 has incompatible type
"str": "Dict[str, str]"; expected "str": "str"  [dict-item]
                    "input": {"label": "first_cat", "input_name": "input1"},
                    ^
test/integration/test_workflow_refactoring.py:157: error: Dict entry 2 has incompatible type
"str": "Dict[str, str]"; expected "str": "str"  [dict-item]
                    "output": {"label": "new_label"},
                    ^
test/integration/test_workflow_refactoring.py:167: error: Dict entry 1 has incompatible type
"str": "Dict[str, str]"; expected "str": "str"  [dict-item]
                    "input": {"label": "first_cat", "input_name": "input1"},
                    ^
test/integration/test_workflow_refactoring.py:168: error: Dict entry 2 has incompatible type
"str": "Dict[str, str]"; expected "str": "str"  [dict-item]
                    "output": {"label": "new_label"},
                    ^
test/integration/test_workflow_refactoring.py:177: error: Dict entry 1 has incompatible type
"str": "Dict[str, str]"; expected "str": "str"  [dict-item]
                    "input": {"label": "first_cat", "input_name": "input1"},
                    ^
test/integration/test_workflow_refactoring.py:188: error: Dict entry 1 has incompatible type
"str": "Dict[str, str]"; expected "str": "str"  [dict-item]
                    "output": {"label": "first_cat", "output_name": "out_file1"},
                    ^
test/integration/test_workflow_refactoring.py:217: error: Dict entry 1 has incompatible type
"str": "List[Dict[str, str]]"; expected "str": "str"  [dict-item]
                {"action_type": "update_creator", "creator": [{"class": "Person", "name...
                                                  ^
test/integration/test_workflow_refactoring.py:226: error: Dict entry 1 has incompatible type
"str": "Dict[str, str]"; expected "str": "str"  [dict-item]
            actions = [{"action_type": "update_report", "report": {"markdown": "my repo...
                                                        ^
test/integration/test_workflow_refactoring.py:235: error: Dict entry 3 has incompatible type
"str": "Dict[str, str]"; expected "str": "str"  [dict-item]
                    "tool_state": {"parameter_type": "text"},
                    ^
test/integration/test_workflow_refactoring.py:236: error: Dict entry 4 has incompatible type
"str": "Dict[str, int]"; expected "str": "str"  [dict-item]
                    "position": {"left": 10, "top": 50},
                    ^
test/integration/test_workflow_refactoring.py:246: error: Dict entry 1 has incompatible type
"str": "Dict[str, str]"; expected "str": "str"  [dict-item]
                    "output": {"label": "first_cat", "output_name": "out_file1"},
                    ^
Found 22 errors in 1 file (checked 1522 source files)
ERROR: InvocationError for command /Users/jxc755/workspace/galaxy/.tox/mypy/bin/mypy test lib (exited with code 1)
```

I don't think typing the helpers is good enough because mypy uses the initial assignment to infer a type instead of how actions is used further along. So it thinks actions is the simplest thing earlier on. Since some of those initial dictionaries are simple and then get more complex - I think it is easier to just to type all the actions.

This way if we use a TypedDict or something later on also we will get even more typing.

Reply:
Ah, didn't spot that some of these `actions` were reused (must have checked one test where it wasn't).